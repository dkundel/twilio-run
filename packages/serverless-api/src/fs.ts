import fs from 'fs';
import path from 'path';
import { promisify } from 'util';

export const access = promisify(fs.access);
export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
export const readdir = promisify(fs.readdir);
export const stat = promisify(fs.stat);

export async function fileExists(filePath: string) {
  try {
    await access(filePath, fs.constants.R_OK | fs.constants.W_OK);
    return true;
  } catch (err) {
    return false;
  }
}

export interface FileInfo {
  name: string;
  path: string;
}

export async function getDirContent(
  dir: string,
  ext?: string
): Promise<FileInfo[]> {
  const rawFiles = await readdir(dir);
  const unfilteredFiles: (FileInfo | undefined)[] = await Promise.all(
    rawFiles.map(async (file: string) => {
      const filePath = path.join(dir, file);
      const entry = await stat(filePath);
      if (!entry.isFile()) {
        return undefined;
      }

      if (ext && path.extname(file) !== ext) {
        return undefined;
      }

      return {
        name: file,
        path: filePath,
      };
    })
  );

  return unfilteredFiles.filter(
    entry => typeof entry !== 'undefined'
  ) as FileInfo[];
}

module.exports = {
  fileExists,
  readFile,
  writeFile,
  readdir,
  getDirContent,
};
