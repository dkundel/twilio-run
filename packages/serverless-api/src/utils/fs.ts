/** @module @twilio-labs/serverless-api/dist/utils/fs */

import debug from 'debug';
import fs from 'fs';
import path from 'path';
import recursiveReadDir from 'recursive-readdir';
import { promisify } from 'util';
import {
  AccessOptions,
  DirectoryContent,
  FileInfo,
  ResourcePathAndAccess,
} from '../types';

const log = debug('twilio-serverless-api:fs');

export const access = promisify(fs.access);
export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
export const readDir = promisify(recursiveReadDir);
export const stat = promisify(fs.stat);

/**
 * Checks if a given file exists by checking if we have read & write access
 *
 * @export
 * @param {string} filePath full path of the file to check
 * @returns
 */
export async function fileExists(filePath: string) {
  try {
    await access(filePath, fs.constants.R_OK | fs.constants.W_OK);
    return true;
  } catch (err) {
    return false;
  }
}

/**
 * Determines the access and Serverless path for a filesystem resource.
 * If it receives an ignore extension it will drop it from the final serverless path
 *
 * @export
 * @param {FileInfo} file the file to get the access and path for
 * @param {string} [ignoreExtension] file extension to drop for serverless path
 * @returns {ResourcePathAndAccess}
 */
export function getPathAndAccessFromFileInfo(
  file: FileInfo,
  ignoreExtension?: string
): ResourcePathAndAccess {
  const relativePath = path.dirname(file.name);

  let access: AccessOptions = 'public';
  const ext = path.extname(file.name);
  let baseName = path.basename(file.name, ext);
  if (file.name.includes(`.protected${ext}`)) {
    access = 'protected';
  } else if (file.name.includes(`.private${ext}`)) {
    access = 'private';
  }
  baseName = baseName.replace(`.${access}`, '');

  let resourcePath = `/` + path.join(relativePath, baseName);
  if (ext !== ignoreExtension) {
    resourcePath += ext;
  }
  resourcePath = resourcePath.replace(/\s/g, '-');

  return {
    path: resourcePath,
    access,
  };
}

/**
 * Retrieves all (nested) files from a given directory.
 *
 * If an extension is specified it will be used to filter the results.
 *
 * @export
 * @param {string} dir the directory to be checked
 * @param {string} [extensionn] extension to be ignored in the results
 * @returns {Promise<FileInfo[]>}
 */
export async function getDirContent(
  dir: string,
  extensionn?: string
): Promise<FileInfo[]> {
  const rawFiles: string[] = (await readDir(dir)) as string[];
  const unfilteredFiles: (FileInfo | undefined)[] = await Promise.all(
    rawFiles.map(async (filePath: string) => {
      if (!path.isAbsolute(filePath)) {
        filePath = path.join(dir, filePath);
      }

      const entry = await stat(filePath);
      if (!entry.isFile()) {
        return undefined;
      }

      if (extensionn && path.extname(filePath) !== extensionn) {
        return undefined;
      }

      const name = path.relative(dir, filePath);
      return {
        name: name,
        path: filePath,
      };
    })
  );

  return unfilteredFiles.filter(
    entry => typeof entry !== 'undefined'
  ) as FileInfo[];
}

/**
 * Given a list of directory names it will return the first one that exists in
 * the base path.
 *
 * **Important**: Performs synchronous file system reading
 *
 * @export
 * @param {string} basePath
 * @param {string[]} directories
 * @returns {string}
 */
export function getFirstMatchingDirectory(
  basePath: string,
  directories: string[]
): string {
  for (let dir of directories) {
    const fullPath = path.join(basePath, dir);

    try {
      const fStat = fs.statSync(fullPath);
      if (fStat.isDirectory()) {
        return fullPath;
      }
    } catch (err) {
      continue;
    }
  }

  throw new Error(
    `Could not find any of these directories "${directories.join('", "')}"`
  );
}

export type SearchConfig = {
  /**
   * Ordered folder names to search for to find functions
   *
   * @type {string[]}
   */
  functionsFolderNames?: string[];
  /**
   * Ordered folder names to search for to find assets
   *
   * @type {string[]}
   */
  assetsFolderNames?: string[];
};

/**
 * Retrieves a list of functions and assets existing in a given base directory
 * Will check for both "functions" and "src" as directory for functions and
 * "assets" and "static" for assets
 *
 * @export
 * @param {string} cwd Directory
 * @param {SearchConfig} config lets you override the folders to use
 * @returns {Promise<DirectoryContent>}
 */
export async function getListOfFunctionsAndAssets(
  cwd: string,
  config: SearchConfig = {}
): Promise<DirectoryContent> {
  console.log(config.assetsFolderNames);
  let functionsDir;
  try {
    const possibleFunctionDirs = config.functionsFolderNames || [
      'functions',
      'src',
    ];
    log('Search for directory. Options: "%s"', possibleFunctionDirs.join(','));
    functionsDir = getFirstMatchingDirectory(cwd, possibleFunctionDirs);
  } catch (err) {
    functionsDir = undefined;
  }

  log('Found Functions Directory "%s"', functionsDir);

  let assetsDir;
  try {
    const possibleAssetDirs = config.assetsFolderNames || ['assets', 'static'];
    log('Search for directory. Options: "%s"', possibleAssetDirs.join(','));
    assetsDir = getFirstMatchingDirectory(cwd, possibleAssetDirs);
  } catch (err) {
    assetsDir = undefined;
  }

  log('Found Assets Direectory "%s"', assetsDir);

  const functions = functionsDir
    ? await getDirContent(functionsDir, '.js')
    : [];
  const assets = assetsDir ? await getDirContent(assetsDir) : [];
  return { functions, assets };
}
