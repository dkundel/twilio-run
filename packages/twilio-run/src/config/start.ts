import { EnvironmentVariables } from '@twilio-labs/serverless-api';
import dotenv from 'dotenv';
import { readFile } from 'fs';
import { promisify } from 'util';
const readFilePromise = promisify(readFile);
import path, { resolve, join } from 'path';
import { homedir } from 'os';
import { Arguments } from 'yargs';
import { ExternalCliOptions, SharedFlags } from '../commands/shared';
import { CliInfo } from '../commands/types';
import { EnvironmentVariablesWithAuth } from '../types/generic';
import { fileExists } from '../utils/fs';
import { getDebugFunction, logger } from '../utils/logger';
import { readSpecializedConfig } from './global';
import { mergeFlagsAndConfig } from './utils/mergeFlagsAndConfig';
import { Ngrok } from 'ngrok';
import { parse } from 'yaml';

const debug = getDebugFunction('twilio-run:cli:config');

type InspectInfo = {
  hostPort: string;
  break: boolean;
};

export type StartCliConfig = {
  inspect?: InspectInfo;
  baseDir: string;
  env: EnvironmentVariablesWithAuth;
  port: number;
  url: string;
  detailedLogs: boolean;
  live: boolean;
  logs: boolean;
  legacyMode: boolean;
  appName: string;
  assetsFolderName?: string;
  functionsFolderName?: string;
  forkProcess: boolean;
};

export type StartCliFlags = Arguments<
  SharedFlags & {
    dir?: string;
    cwd?: string;
    loadLocalEnv: boolean;
    env?: string;
    port: string;
    ngrok?: string | boolean;
    ngrokConfig?: string;
    ngrokName?: string;
    logs: boolean;
    detailedLogs: boolean;
    live: boolean;
    inspect?: string;
    inspectBrk?: string;
    legacyMode: boolean;
    assetsFolder?: string;
    functionsFolder?: string;
    experimentalForkProcess: boolean;
  }
>;

export type WrappedStartCliFlags = {
  flags: StartCliFlags;
};

export async function getUrl(cli: StartCliFlags, port: string | number) {
  let url = `http://localhost:${port}`;
  if (typeof cli.ngrok !== 'undefined') {
    debug('Starting ngrok tunnel');
    // Setup default ngrok config, setting the protocol and the port number to
    // forward to.
    const defaultConfig: Ngrok.Options = { addr: port, proto: 'http' };
    let tunnelConfig = defaultConfig;
    let ngrokConfig;
    if (typeof cli.ngrokConfig === 'string') {
      // If we set a config path then try to load that config. If the config
      // fails to load then we'll try to load the default config instead.
      const configPath = join(cli.cwd || process.cwd(), cli.ngrokConfig);
      try {
        ngrokConfig = parse(await readFilePromise(configPath, 'utf-8'));
      } catch (err) {
        logger.warn(`Could not find ngrok config file at ${configPath}`);
      }
    }
    if (!ngrokConfig) {
      // Try to load default config. If there is no default config file, set
      // `ngrokConfig` to be an empty object.
      const configPath = join(homedir(), '.ngrok2', 'ngrok.yml');
      try {
        ngrokConfig = parse(await readFilePromise(configPath, 'utf-8'));
      } catch (err) {
        ngrokConfig = {};
      }
    }
    if (
      typeof cli.ngrokName === 'string' &&
      typeof ngrokConfig.tunnels === 'object'
    ) {
      // If we've asked for a named ngrok tunnel and there are available tunnels
      // in the config, then set the `tunnelConfig` to the options from the
      // config, overriding the addr and proto to the defaults.
      tunnelConfig = { ...ngrokConfig.tunnels[cli.ngrokName], ...tunnelConfig };
      if (!tunnelConfig) {
        // If the config does not include the named tunnel, then set it back to
        // the default options.
        logger.warn(
          `Could not find config for named tunnel "${cli.ngrokName}". Falling back to other options.`
        );
        tunnelConfig = defaultConfig;
      }
    }
    if (typeof ngrokConfig.authtoken === 'string') {
      // If there is an authtoken in the config, add it to the tunnel config.
      tunnelConfig.authToken = ngrokConfig.authtoken;
    }
    if (typeof cli.ngrok === 'string' && cli.ngrok.length > 0) {
      // If we've asked for a custom subdomain, override the tunnel config with
      // it.
      tunnelConfig.subdomain = cli.ngrok;
    }
    const ngrok = require('ngrok');
    try {
      // Try to open the ngrok tunnel.
      url = await ngrok.connect(tunnelConfig);
    } catch (error) {
      // If it fails, it is likely to be because the tunnel config we pass is
      // not allowed (e.g. using a custom subdomain without an authtoken). The
      // error message from ngrok itself should describe the issue.
      logger.warn(error.message);
      if (
        typeof error.details !== 'undefined' &&
        typeof error.details.err !== 'undefined'
      ) {
        logger.warn(error.details.err);
      }
    }
    debug('ngrok tunnel URL: %s', url);
  }

  return url;
}

export function getPort(cli: StartCliFlags): number {
  let port = process.env.PORT || 3000;
  if (typeof cli.port !== 'undefined') {
    port = parseInt(cli.port, 10);
    debug('Overriding port via command-line flag to %d', port);
  }
  if (typeof port === 'string') {
    port = parseInt(port, 10);
  }
  return port;
}

export async function getEnvironment(
  cli: StartCliFlags,
  baseDir: string
): Promise<EnvironmentVariables> {
  let env: EnvironmentVariables = {};
  if (cli.loadLocalEnv) {
    debug('Loading local environment variables');
    env = { ...process.env };
  }

  const envFilePath = cli.env || '.env';
  const fullEnvPath = resolve(baseDir || process.cwd(), envFilePath);
  if (await fileExists(fullEnvPath)) {
    try {
      debug(`Read .env file at "%s"`, fullEnvPath);
      const envContent = await readFilePromise(fullEnvPath, 'utf8');
      const envValues = dotenv.parse(envContent);
      for (const [key, val] of Object.entries(envValues)) {
        env[key] = val;
      }
    } catch (err) {
      logger.error('Failed to read .env file');
    }
  } else {
    if (cli.env) {
      logger.error('Failed to find .env file');
    }
    debug('Not loading a .env file');
  }
  return env;
}

export function getBaseDirectory(cli: StartCliFlags): string {
  let baseDir = process.cwd();
  if (cli.cwd) {
    baseDir = path.resolve(cli.cwd);
    debug('Set base directory based on input to "%s"', baseDir);
  } else if (cli.dir) {
    baseDir = path.resolve(cli.dir);
    debug('Set base directory based on input to "%s"', baseDir);
  }
  return baseDir;
}

export function getInspectInfo(cli: StartCliFlags): InspectInfo | undefined {
  if (typeof cli.inspectBrk !== 'undefined') {
    return {
      hostPort: cli.inspectBrk,
      break: true,
    };
  } else if (typeof cli.inspect !== 'undefined') {
    return { hostPort: cli.inspect, break: false };
  }
  return undefined;
}

export async function getConfigFromCli(
  flags: StartCliFlags,
  cliInfo: CliInfo = { options: {} },
  externalCliOptions?: ExternalCliOptions
): Promise<StartCliConfig> {
  const configFlags = readSpecializedConfig(
    flags.cwd || process.cwd(),
    flags.config,
    'startConfig',
    {
      projectId:
        (externalCliOptions && externalCliOptions.accountSid) || undefined,
    }
  ) as StartCliFlags;
  const cli = mergeFlagsAndConfig(configFlags, flags, cliInfo);
  const config = {} as StartCliConfig;

  config.inspect = getInspectInfo(cli);
  config.baseDir = getBaseDirectory(cli);
  config.env = await getEnvironment(cli, config.baseDir);
  config.port = getPort(cli);
  config.detailedLogs = cli.detailedLogs;
  config.live = cli.live;
  config.logs = cli.logs;
  config.legacyMode = cli.legacyMode;
  config.appName = 'twilio-run';
  config.assetsFolderName = cli.assetsFolder;
  config.functionsFolderName = cli.functionsFolder;
  config.forkProcess = cli.experimentalForkProcess;

  return config;
}
