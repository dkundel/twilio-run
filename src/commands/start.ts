import inquirer from 'inquirer';
import { Argv } from 'yargs';
import checkNodejsVersion from '../checks/nodejs-version';
import checkProjectStructure from '../checks/project-structure';
import { getConfigFromCli, StartCliFlags } from '../config/start';
import { printRouteInfo } from '../printers/start';
import { createServer } from '../runtime/server';
import { startInspector } from '../runtime/utils/inspector';
import { getDebugFunction, logger, setLogLevelByName } from '../utils/logger';
import { ExternalCliOptions, sharedCliOptions } from './shared';
import { CliInfo } from './types';
import { getFullCommand } from './utils';

const debug = getDebugFunction('twilio-run:start');

type ServerError = Error & {
  code: string;
};

function randomPort() {
  // Returns a random port number higher than 1024 and lower than 65536.
  return Math.floor(Math.random() * (65535 - 1025) + 1025);
}

function validatePortNumber(input: string) {
  const newPortNumber = parseInt(input, 10);
  if (
    !Number.isNaN(newPortNumber) &&
    newPortNumber <= 65535 &&
    newPortNumber > 1024
  ) {
    return true;
  }
  return 'Please enter a port number between 1025 and 65535.';
}

export async function handler(
  argv: StartCliFlags,
  externalCliOptions?: ExternalCliOptions
): Promise<void> {
  setLogLevelByName(argv.logLevel);

  checkNodejsVersion();

  const config = await getConfigFromCli(argv, cliInfo, externalCliOptions);

  const command = getFullCommand(argv);
  const directories = {
    assetsDirectories: config.assetsFolderName
      ? [config.assetsFolderName]
      : undefined,
    functionsDirectories: config.functionsFolderName
      ? [config.functionsFolderName]
      : undefined,
  };
  await checkProjectStructure(config.baseDir, command, false, directories);

  debug('Determined configuration: %p', config);
  process.title = config.appName;

  debug('Set environment variables as: %r', config.env);
  process.env = config.env;

  if (config.inspect) {
    debug(
      'Starting inspector mode with following configuration: %p',
      config.inspect
    );
    startInspector(config.inspect.hostPort, config.inspect.break);
  }

  const app = await createServer(config.port, config);
  debug('Start server on port %d', config.port);
  return new Promise((resolve, reject) => {
    let attempts = 1;
    const MAX_ATTEMPTS = 3;
    const serverStartedSuccessfully = async () => {
      printRouteInfo(config);
      resolve();
    };
    const handleServerError = async (error: ServerError) => {
      if (error.code === 'EADDRINUSE') {
        if (attempts > MAX_ATTEMPTS) {
          logger.info('Too many retries. Please check your available ports.');
          process.exit(1);
        } else {
          const answers = await inquirer.prompt([
            {
              type: 'input',
              default: randomPort(),
              name: 'newPortNumber',
              message: `Port ${config.port} is already in use. Choose a new port number:`,
              validate: validatePortNumber,
            },
          ]);
          attempts += 1;
          const server = app.listen(
            answers.newPortNumber,
            serverStartedSuccessfully
          );
          server.on('error', handleServerError);
        }
      } else {
        reject(error);
      }
    };

    const server = app.listen(config.port, serverStartedSuccessfully);
    server.on('error', handleServerError);
  });
}

export const cliInfo: CliInfo = {
  options: {
    ...sharedCliOptions,
    'load-local-env': {
      alias: 'f',
      default: false,
      type: 'boolean',
      describe: 'Includes the local environment variables',
    },
    cwd: {
      type: 'string',
      describe:
        'Alternative way to define the directory to start the server in. Overrides the [dir] argument passed.',
    },
    env: {
      alias: 'e',
      type: 'string',
      describe: 'Loads .env file, overrides local env variables',
    },
    port: {
      alias: 'p',
      type: 'string',
      describe: 'Override default port of 3000',
      default: '3000',
      requiresArg: true,
    },
    ngrok: {
      type: 'string',
      describe: 'Uses ngrok to create and outfacing url',
    },
    logs: {
      type: 'boolean',
      default: true,
      describe: 'Toggles request logging',
    },
    'detailed-logs': {
      type: 'boolean',
      default: false,
      describe:
        'Toggles detailed request logging by showing request body and query params',
    },
    live: {
      type: 'boolean',
      default: true,
      describe: 'Always serve from the current functions (no caching)',
    },
    inspect: {
      type: 'string',
      describe: 'Enables Node.js debugging protocol',
    },
    'inspect-brk': {
      type: 'string',
      describe:
        'Enables Node.js debugging protocol, stops executioin until debugger is attached',
    },
    'legacy-mode': {
      type: 'boolean',
      describe:
        'Enables legacy mode, it will prefix your asset paths with /assets',
    },
    'assets-folder': {
      type: 'string',
      describe: 'Specific folder name to be used for static assets',
    },
    'functions-folder': {
      type: 'string',
      describe: 'Specific folder name to be used for static functions',
    },
  },
};

function optionBuilder(yargs: Argv<any>): Argv<StartCliFlags> {
  yargs = yargs
    .example('$0', 'Serves all functions in current functions subdirectory')
    .example('$0 start demo', 'Serves all functions in demo/functions')
    .example('PORT=9000 $0', 'Serves functions on port 9000')
    .example('$0 start --port=4200', 'Serves functions on port 4200')
    .example('$0 start --env', 'Loads environment variables from .env file')
    .example(
      '$0 start --ngrok',
      'Exposes the Twilio functions via ngrok to share them'
    )
    .epilog('for more information, check out https://twil.io/local-functions');

  yargs = Object.keys(cliInfo.options).reduce((yargs, name) => {
    return yargs.option(name, cliInfo.options[name]);
  }, yargs);

  return yargs;
}

export const command = ['start [dir]', '$0 [dir]'];
export const describe = 'Starts local Twilio Functions development server';
export const builder = optionBuilder;
