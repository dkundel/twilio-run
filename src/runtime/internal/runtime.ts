import {
  AssetResourceMap,
  ResourceMap,
  RuntimeInstance,
  RuntimeSyncClientOptions,
  RuntimeSyncServiceContext,
} from '@twilio-labs/serverless-runtime-types/types';
import debug from 'debug';
import { readFileSync } from 'fs';
import twilio from 'twilio';
import { ServiceContext } from 'twilio/lib/rest/sync/v1/service';
import { SyncListListInstance } from 'twilio/lib/rest/sync/v1/service/syncList';
import { SyncMapListInstance } from 'twilio/lib/rest/sync/v1/service/syncMap';
import { StartCliConfig } from '../../config/start';

const log = debug('twilio-run:runtime');

const { getCachedResources } = require('./route-cache');

function getAssets(): AssetResourceMap {
  const { assets } = getCachedResources();
  if (assets.length === 0) {
    return {};
  }

  const result: AssetResourceMap = {};
  for (const asset of assets) {
    if (asset.access === 'private') {
      const prefix =
        process.env.TWILIO_FUNCTIONS_LEGACY_MODE === 'true' ? '/assets' : '';
      const open = () => readFileSync(asset.path, 'utf8');
      result[prefix + asset.assetPath] = { path: asset.path, open };
    }
  }
  log('Found the following assets available: %O', result);
  return result;
}

function getFunctions(): ResourceMap {
  const { functions } = getCachedResources();
  if (functions.length === 0) {
    return {};
  }

  const result: ResourceMap = {};
  for (const fn of functions) {
    result[fn.functionPath.substr(1)] = { path: fn.path };
  }
  log('Found the following functions available: %O', result);
  return result;
}

export type ExtendedSyncServiceContext = ServiceContext & {
  maps: SyncMapListInstance;
  lists: SyncListListInstance;
};

export function create({ env }: StartCliConfig): RuntimeInstance {
  function getSync(
    options?: RuntimeSyncClientOptions
  ): RuntimeSyncServiceContext {
    options = { serviceName: 'default', ...options };
    const { serviceName } = options;
    delete options.serviceName;

    const client = twilio(env.ACCOUNT_SID, env.AUTH_TOKEN, options);
    const service = client.sync.services(
      serviceName || 'default'
    ) as RuntimeSyncServiceContext;

    service.maps = service.syncMaps;
    service.lists = service.syncLists;
    return service;
  }

  return { getSync, getAssets, getFunctions };
}

module.exports = { create };
