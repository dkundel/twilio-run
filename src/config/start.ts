import { EnvironmentVariables } from '@twilio-labs/serverless-api';
import debugModule from 'debug';
import dotenv from 'dotenv';
import { readFileSync } from 'fs';
import logSymbols from 'log-symbols';
import path, { resolve } from 'path';
import { Arguments } from 'yargs';
import { ExternalCliOptions, SharedFlags } from '../commands/shared';
import { CliInfo } from '../commands/types';
import { EnvironmentVariablesWithAuth } from '../types/generic';
import { fileExists } from '../utils/fs';
import { readSpecializedConfig } from './global';
import { mergeFlagsAndConfig } from './utils/mergeFlagsAndConfig';

const debug = debugModule('twilio-run:cli:config');

type NgrokConfig = {
  addr: string | number;
  subdomain?: string;
};

type InspectInfo = {
  hostPort: string;
  break: boolean;
};

export type StartCliConfig = {
  inspect?: InspectInfo;
  baseDir: string;
  env: EnvironmentVariablesWithAuth;
  port: number;
  url: string;
  detailedLogs: boolean;
  live: boolean;
  logs: boolean;
  legacyMode: boolean;
  appName: string;
};

export type StartCliFlags = Arguments<
  SharedFlags & {
    dir?: string;
    cwd?: string;
    loadLocalEnv: boolean;
    env?: string;
    port: string;
    ngrok?: string | boolean;
    logs: boolean;
    detailedLogs: boolean;
    live: boolean;
    inspect?: string;
    inspectBrk?: string;
    legacyMode: boolean;
  }
>;

export type WrappedStartCliFlags = {
  flags: StartCliFlags;
};

export async function getUrl(cli: StartCliFlags, port: string | number) {
  let url = `http://localhost:${port}`;
  if (typeof cli.ngrok !== 'undefined') {
    debug('Starting ngrok tunnel');
    const ngrokConfig: NgrokConfig = { addr: port };
    if (typeof cli.ngrok === 'string' && cli.ngrok.length > 0) {
      ngrokConfig.subdomain = cli.ngrok;
    }

    url = await require('ngrok').connect(ngrokConfig);
    debug('ngrok tunnel URL: %s', url);
  }

  return url;
}

export function getPort(cli: StartCliFlags): number {
  let port = process.env.PORT || 3000;
  if (typeof cli.port !== 'undefined') {
    port = parseInt(cli.port, 10);
    debug('Overriding port via command-line flag to %d', port);
  }
  if (typeof port === 'string') {
    port = parseInt(port, 10);
  }
  return port;
}

export async function getEnvironment(
  cli: StartCliFlags,
  baseDir: string
): Promise<EnvironmentVariables> {
  let env: EnvironmentVariables = {};
  if (cli.loadLocalEnv) {
    debug('Loading local environment variables');
    env = { ...process.env };
  }

  const envFilePath = cli.env || '.env';
  const fullEnvPath = resolve(baseDir || process.cwd(), envFilePath);
  if (await fileExists(fullEnvPath)) {
    try {
      debug(`Read .env file at "%s"`, fullEnvPath);
      const envContent = readFileSync(fullEnvPath, 'utf8');
      const envValues = dotenv.parse(envContent);
      for (const [key, val] of Object.entries(envValues)) {
        env[key] = val;
      }
    } catch (err) {
      console.error(logSymbols.error, 'Failed to read .env file');
    }
  } else {
    if (cli.env) {
      console.error(logSymbols.error, 'Failed to find .env file');
    }
    debug('Not loading a .env file');
  }
  return env;
}

export function getBaseDirectory(cli: StartCliFlags): string {
  let baseDir = process.cwd();
  if (cli.cwd) {
    baseDir = path.resolve(cli.cwd);
    debug('Set base directory based on input to "%s"', baseDir);
  } else if (cli.dir) {
    baseDir = path.resolve(cli.dir);
    debug('Set base directory based on input to "%s"', baseDir);
  }
  return baseDir;
}

export function getInspectInfo(cli: StartCliFlags): InspectInfo | undefined {
  if (typeof cli.inspectBrk !== 'undefined') {
    return {
      hostPort: cli.inspectBrk,
      break: true,
    };
  } else if (typeof cli.inspect !== 'undefined') {
    return { hostPort: cli.inspect, break: false };
  }
  return undefined;
}

export async function getConfigFromCli(
  flags: StartCliFlags,
  cliInfo: CliInfo = { options: {} },
  externalCliOptions?: ExternalCliOptions
): Promise<StartCliConfig> {
  const configFlags = readSpecializedConfig(
    flags.cwd || process.cwd(),
    flags.config,
    'startConfig',
    {
      projectId:
        (externalCliOptions && externalCliOptions.accountSid) || undefined,
    }
  ) as StartCliFlags;
  const cli = mergeFlagsAndConfig(configFlags, flags, cliInfo);
  const config = {} as StartCliConfig;

  config.inspect = getInspectInfo(cli);
  config.baseDir = getBaseDirectory(cli);
  config.env = await getEnvironment(cli, config.baseDir);
  config.port = getPort(cli);
  config.url = await getUrl(cli, config.port);
  config.detailedLogs = cli.detailedLogs;
  config.live = cli.live;
  config.logs = cli.logs;
  config.legacyMode = cli.legacyMode;
  config.appName = 'twilio-run';

  return config;
}
