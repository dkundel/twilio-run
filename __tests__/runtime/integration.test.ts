jest.unmock('twilio');

import { Express } from 'express';
import { readdirSync } from 'fs';
import { basename, resolve } from 'path';
import request from 'supertest';
import { StartCliConfig } from '../../src/config/start';
import { createServer } from '../../src/runtime/server';

const TEST_DIR = resolve(__dirname, '../../fixtures');

const TEST_FUNCTIONS_DIR = resolve(TEST_DIR, 'functions');
const TEST_ASSETS_DIR = resolve(TEST_DIR, 'assets');
const TEST_ENV = {};

const availableFunctions = readdirSync(TEST_FUNCTIONS_DIR).map(
  (name: string) => {
    const path = resolve(TEST_FUNCTIONS_DIR, name);
    const url = `/${basename(name, '.js')}`;
    return { name, url, path };
  }
);
const availableAssets = readdirSync(TEST_ASSETS_DIR).map((name: string) => {
  const path = resolve(TEST_ASSETS_DIR, name);
  const url = `/${name}`;
  return { name, url, path };
});

type InternalResponse = request.Response & {
  statusCode: number;
  headers: {
    [key: string]: string | undefined;
  };
};

function responseToSnapshotJson(response: InternalResponse) {
  let { statusCode, type, body, text, headers } = response;
  delete headers['date'];
  delete headers['last-modified'];

  if (text && text.startsWith('Error')) {
    // stack traces are different in every environment
    // let's not snapshot values that rely on it
    text = `${text.split('\n')[0]} ...`;
  }
  delete headers['content-length'];
  delete headers['etag'];

  return {
    statusCode,
    type,
    body,
    text,
    headers,
  };
}

describe('with an express app', () => {
  let app: Express;

  describe('with inline function handling', () => {
    beforeAll(async () => {
      app = await createServer(9000, {
        baseDir: TEST_DIR,
        env: TEST_ENV,
        logs: false,
      } as StartCliConfig);
    });

    describe('Function integration tests', () => {
      for (const testFnCode of availableFunctions) {
        test(`${testFnCode.name} should match snapshot`, async () => {
          const response = await request(app).get(testFnCode.url);
          if (response.status === 500) {
            expect(response.text).toMatch(/Error/);
          } else {
            const result = responseToSnapshotJson(response as InternalResponse);
            expect(result).toMatchSnapshot();
          }
        });
      }
    });

    describe('Assets integration tests', () => {
      for (const testAsset of availableAssets) {
        test(`${testAsset.name} should match snapshot`, async () => {
          const response = await request(app).get(testAsset.url);
          const result = responseToSnapshotJson(response as InternalResponse);
          expect(result).toMatchSnapshot();
        });

        test(`OPTIONS request to ${testAsset.name} should return CORS headers and no body`, async () => {
          const response = (await request(app).options(
            testAsset.url
          )) as InternalResponse;
          expect(response.headers['access-control-allow-origin']).toEqual('*');
          expect(response.headers['access-control-allow-headers']).toEqual(
            'Accept, Authorization, Content-Type, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since'
          );
          expect(response.headers['access-control-allow-methods']).toEqual(
            'GET, POST, OPTIONS'
          );
          expect(response.headers['access-control-expose-headers']).toEqual(
            'ETag'
          );
          expect(response.headers['access-control-max-age']).toEqual('86400');
          expect(response.headers['access-control-allow-credentials']).toEqual(
            'true'
          );
          expect(response.text).toEqual('');
        });

        test(`GET request to ${testAsset.name} should not return CORS headers`, async () => {
          const response = (await request(app).get(
            testAsset.url
          )) as InternalResponse;
          expect(response.headers['access-control-allow-origin']).toBeUndefined();
          expect(
            response.headers['access-control-allow-headers']
          ).toBeUndefined();
          expect(
            response.headers['access-control-allow-methods']
          ).toBeUndefined();
          expect(
            response.headers['access-control-expose-headers']
          ).toBeUndefined();
          expect(response.headers['access-control-max-age']).toBeUndefined();
          expect(
            response.headers['access-control-allow-credentials']
          ).toBeUndefined();
        });
      }
    });
  });
  describe('with forked process function handling', () => {
    beforeAll(async () => {
      app = await createServer(9000, {
        baseDir: TEST_DIR,
        env: TEST_ENV,
        logs: false,
        forkProcess: true
      } as StartCliConfig);
    });

    describe('Function integration tests', () => {
      for (const testFnCode of availableFunctions) {
        test(`${testFnCode.name} should match snapshot`, async () => {
          const response = await request(app).get(testFnCode.url);
          if (response.status === 500) {
            expect(response.text).toMatch(/Error/);
          } else {
            const result = responseToSnapshotJson(response as InternalResponse);
            expect(result).toMatchSnapshot();
          }
        });
      }
    });
  });
});
