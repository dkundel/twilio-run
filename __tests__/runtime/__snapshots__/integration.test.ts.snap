// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with an express app Assets integration tests hello.js should match snapshot 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "accept-ranges": "bytes",
    "cache-control": "public, max-age=0",
    "connection": "close",
    "content-type": "application/javascript; charset=UTF-8",
    "last-modified": "Tue, 09 Jun 2020 01:11:06 GMT",
    "x-powered-by": "Express",
  },
  "statusCode": 200,
  "text": "alert('Hello world!');
",
  "type": "application/javascript",
}
`;

exports[`with an express app Function integration tests basic-twiml.js should match snapshot 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "connection": "close",
    "content-type": "text/xml; charset=utf-8",
    "x-powered-by": "Express",
  },
  "statusCode": 200,
  "text": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><Response><Message>Hello World</Message></Response>",
  "type": "text/xml",
}
`;
